### DISCUSSION TITLE
Are binary installs supported?

### DISCUSSION THREAD

**Discussion by leetrout** [astral-sh/rye#1228] [Q&A]
I tried to add the `psycopg2` binary and it failed trying to build the source. My expectation is that it would install the pre-compiled library.

```
rye add psycopg2-binary
```

```
Added psycopg2-binary>=2.9.9 as regular dependency
Reusing already existing virtualenv
Generating production lockfile: /home/user/code/github.com/leetrout/ddl-shenanigans/requirements.lock
Generating dev lockfile: /home/user/code/github.com/leetrout/ddl-shenanigans/requirements-dev.lock
Installing dependencies
Resolved 15 packages in 19ms
   Built ddl-shenanigans @ file:///home/user/code/github.com/leetrout/ddl-shenanigans
error: Failed to prepare distributions
  Caused by: Failed to fetch wheel: psycopg2==2.9.9
  Caused by: Failed to build: `psycopg2==2.9.9`
  Caused by: Build backend failed to build wheel through `build_wheel()` with exit status: 1
--- stdout:
running bdist_wheel
running build
running build_py
copying lib/pool.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/_ipaddress.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/__init__.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/extensions.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/sql.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/extras.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/errorcodes.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/tz.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/_json.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/_range.py -> build/lib.linux-x86_64-cpython-312/psycopg2
copying lib/errors.py -> build/lib.linux-x86_64-cpython-312/psycopg2
running build_ext
building 'psycopg2._psycopg' extension
clang -pthread -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -fPIC -fPIC "-DPSYCOPG_VERSION=2.9.9 (dt dec pq3 ext lo64)" -DPSYCOPG_DEBUG=1 -DPG_VERSION_NUM=140012 -DHAVE_LO64=1 -DPSYCOPG_DEBUG=1 -I/home/user/.cache/uv/builds-v0/.tmpJ3m2dz/include -I/home/user/.rye/py/cpython@3.12.3/include/python3.12 -I. -I/usr/include/postgresql -I/usr/include/postgresql/14/server -I/usr/include/libxml2 -c psycopg/adapter_asis.c -o build/temp.linux-x86_64-cpython-312/psycopg/adapter_asis.o -Wdeclaration-after-statement
--- stderr:
It appears you are missing some prerequisite to build the package from source.

You may install a binary package by installing 'psycopg2-binary' from PyPI.
If you want to install psycopg2 from source, please install the packages
required for the build and try again.

For further information please check the 'doc/src/install.rst' file (also at
<https://www.psycopg.org/docs/install.html>).

error: command 'clang' failed: No such file or directory
---
error: Installation of dependencies failed in venv at /home/user/code/github.com/leetrout/ddl-shenanigans/.venv. uv exited with status: exit status: 2
```

**Comment by bluss** [comment]
Of course rye and uv prefer using binary wheels if they are available.

There could be several reasons they are not - a wheel is probably not available for the chosen python version and platform. Sometimes a python version that is more contemporary with the package's release is better (is there a wheel for Python 3.11 rather than 3.12, things like that?)


In this case it falls back to build from source. Failure to build due to the bug https://github.com/astral-sh/rye/issues/646 I think.

One workaround could be something as simple as installing clang or gcc and/or retrying with CC=gcc set (so that it picks a different compiler, since clang was missing according to the error message.) This workaround doesn't always work.

**Reply by bluss** [reply]
There is a wheel for linux x86-64 and cpython 3.12 and rye+uv can find it:

```
DEBUG Selecting: psycopg2-binary==2.9.9 (psycopg2_binary-2.9.9-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl)
DEBUG No cache entry for: https://files.pythonhosted.org/packages/19/57/9f172b900795ea37246c78b5f52e00f4779984370855b3e161600156906d/psycopg2_binary-2.9.9-cp312-cp312-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata
```


A log from `UV_NO_CACHE=1 rye sync -f -v` or similar could be helpful to know why it's not finding the same thing for you(?) maybe.

**Comment by cici** [comment]
I'm running into the same issue.  I get this error for both trying to add psycopg2 or psycopg2-binary:

error: Failed to prepare distributions
  Caused by: Failed to fetch wheel: psycopg2==2.9.9
  Caused by: Failed to build: `psycopg2==2.9.9`
  Caused by: Build backend failed to determine extra requires with `build_wheel()` with exit status: 1

**Comment by ernovout** [comment]
:point_up: Ran into the same issue for both psycopg2 and psycopg2-binary. 

For me the install attempt failed with 
```
error: command 'clang' failed: No such file or directory
```
Installed clang with `apt install clang` and tried `rye add psycopg2` again.

The next attempt failed with a slightly different output:
```
...
./psycopg/psycopg.h:36:10: fatal error: 'libpq-fe.h' file not found
...
error: command '/usr/bin/clang' failed with exit code 1
```

Apparently `libpq-fe.h` is a development header so I also had to `apt install libpq-dev`

After installing these two dependencies I was finally able to add psycopg2 successfully :tada: 

I hope this helps :slightly_smiling_face: Remember to swap `apt` for the os package manager at your disposal.

**Comment by Mahdi-Hosseinali** [comment]
I ran into this error with `mysqlclient`, using `CC=gcc` worked but it would be nice if rye could try using GCC automatically when it cannot find clang.
